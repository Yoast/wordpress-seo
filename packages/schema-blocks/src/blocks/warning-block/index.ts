import { WarningBlockAttributes } from "./warningBlockAttributes";

export { WarningBlockAttributes };
