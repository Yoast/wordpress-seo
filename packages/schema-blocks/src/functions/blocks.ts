import { select } from "@wordpress/data";

/**
 * Gets attributes of inner blocks.
 *
 * @param clientId The client ID of the parent block.
 * @param blocks   A mapping of block name to attribute key.
 *
 * @returns An array contain block names and values.
 */
export function getInnerBlocksAttributes( clientId: string, blocks: Record<string, string> ): { name: string; value: unknown }[] {
	let innerBlocks = select( "core/block-editor" ).getBlock( clientId ).innerBlocks;
	innerBlocks     = innerBlocks.filter( block => block.name in blocks );

	const values = [];

	for ( const block of innerBlocks ) {
		const key   = blocks[ block.name ];
		values.push( { name: block.name, value: block.attributes[ key ] } );
	}

	return values;
}
