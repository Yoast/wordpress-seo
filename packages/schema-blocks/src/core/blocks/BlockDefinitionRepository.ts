import BlockDefinition from "./BlockDefinition";

// Internal store of all known BlockDefinitions.
const registeredBlockDefinitions: Record<string, BlockDefinition> = {};

/**
 * Finds a registered BlockDefinition.
 *
 * @param name The Block to search for.
 *
 * @returns {BlockDefinition} The found block.
 */
export function getBlockDefinition( name: string ): BlockDefinition {
	return registeredBlockDefinitions[ name ];
}

/**
 * Registers a BlockDefinition.
 *
 * @param name       The name of the Definition to register.
 * @param definition The Definition to register.
 */
export function registerBlockDefinition( name: string, definition: BlockDefinition ): void {
	registeredBlockDefinitions[ name ] = definition;
}
