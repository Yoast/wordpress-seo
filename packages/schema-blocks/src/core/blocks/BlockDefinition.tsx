import { ReactElement } from "react";
import { createElement, Fragment } from "@wordpress/element";
import {
	registerBlockType,
	BlockConfiguration,
	BlockEditProps,
	BlockSaveProps,
} from "@wordpress/blocks";
import { InspectorControls, BlockIcon } from "@wordpress/block-editor";
import { select } from "@wordpress/data";
import { PanelBody } from "@wordpress/components";

import BlockInstruction from "./BlockInstruction";
import Definition from "../Definition";
import BlockRootLeaf from "../../leaves/blocks/BlockRootLeaf";
import parse from "../../functions/blocks/parse";
import { registerBlockDefinition } from "./BlockDefinitionRepository";
import logger from "../../functions/logger";
import { openGeneralSidebar } from "../../functions/gutenberg/sidebar";

export interface RenderEditProps extends BlockEditProps<Record<string, unknown>> {
	clientId: string;
	name?: string;
}

export interface RenderSaveProps extends BlockSaveProps<Record<string, unknown>> {
	clientId?: string;
	className?: string;
}

export type MutableBlockConfiguration = {
	-readonly [K in keyof BlockConfiguration]: BlockConfiguration[K]
}

/**
 * BlockDefinition class.
 */
export default class BlockDefinition extends Definition {
	public static separatorCharacters = [ "b", "c", "d", "f", "g", "h", "k", "m", "z" ];
	public static parser = parse;

	public instructions: Record<string, BlockInstruction>;
	public tree: BlockRootLeaf;

	/**
	 * Renders editing the block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	edit( props: RenderEditProps ): ReactElement {
		// Force the sidebar open.
		if ( select( "core/block-editor" ).isBlockSelected( props.clientId ) ) {
			openGeneralSidebar( "edit-post/block", true );
		}

		const sidebarElements = this.sidebarElements( props );

		// Take the children directly to avoid creating too many Fragments.
		const elements = this.tree.children.map( ( leaf, i ) => leaf.edit( props, i ) ).filter( e => e !== null );

		if ( sidebarElements.length > 0 ) {
			const sidebarContainer =
				<InspectorControls key="sidebar-inspector-controls">
					<PanelBody>
						{ sidebarElements }
					</PanelBody>
				</InspectorControls>;

			elements.unshift( sidebarContainer );
		}

		if ( elements.length === 1 ) {
			return elements[ 0 ] as ReactElement;
		}

		return createElement( Fragment, { key: props.clientId }, elements.map( ( element, i ) => createElement( Fragment, { key: i }, element ) ) );
	}

	/**
	 * Renders the persisted block.
	 *
	 * @param props The props.
	 *
	 * @returns The rendered block.
	 */
	save( props: RenderSaveProps ): ReactElement {
		return this.tree.save( props );
	}

	/**
	 * Registers the block with Gutenberg.
	 */
	register(): void {
		const configuration = this.configuration() as MutableBlockConfiguration;
		const name = configuration.name as string;
		delete configuration.name;

		configuration.edit = props => this.edit( props );
		configuration.save = props => this.save( props );

		logger.debug( "registering block " + name );

		if ( configuration.icon && typeof configuration.icon === "string" && configuration.icon.startsWith( "<svg" ) ) {
			configuration.icon = this.createBlockIcon( configuration );
		}

		// Register the block to WordPress.
		registerBlockType( name, configuration );
		// Register the block with our own code.
		registerBlockDefinition( name, this );
	}

	/**
	 * Creates the sidebar elements.
	 *
	 * @param props The properties of the block to create a sidebar for.
	 *
	 * @returns The sidebar element to render.
	 */
	sidebarElements( props: RenderEditProps ): ReactElement[] {
		return Object.values( this.instructions )
			.map( ( instruction, index ) => <Fragment key={ instruction.id }>
				{ instruction.sidebar( props, index ) }
			</Fragment> )
			.filter( e => e !== null );
	}

	/**
	 * Creates a block icon.
	 *
	 * @param configuration The block configuration.
	 *
	 * @returns The sidebar element to render.
	 */
	private createBlockIcon( configuration: MutableBlockConfiguration ): ReactElement {
		const icon = <span
			className="yoast-schema-blocks-icon"
			dangerouslySetInnerHTML={ { __html: configuration.icon as string } }
		/>;

		return <BlockIcon icon={ icon } />;
	}
}
