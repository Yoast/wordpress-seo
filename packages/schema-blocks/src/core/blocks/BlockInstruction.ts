import BlockLeaf from "./BlockLeaf";
import { RenderEditProps, RenderSaveProps } from "./BlockDefinition";
import { ReactElement } from "react";
import { BlockConfiguration, BlockInstance } from "@wordpress/blocks";
import { BlockValidation, BlockValidationResult, BlockPresence } from "../validation";
import Instruction from "../Instruction";

/**
 * BlockInstruction class.
 */
export default abstract class BlockInstruction extends Instruction {
	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders saving the element.
	 *
	 * @param props The props.
	 * @param leaf  The leaf being rendered.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The element to render.
	 */
	save( props: RenderSaveProps, leaf: BlockLeaf, i: number ): ReactElement | string {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders editing the element.
	 *
	 * @param props The props.
	 * @param leaf  The leaf being rendered.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The element to render.
	 */
	edit( props: RenderEditProps, leaf: BlockLeaf, i: number ): ReactElement | string {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/* eslint-disable @typescript-eslint/no-unused-vars */
	/**
	 * Renders the sidebar.
	 *
	 * @param props The props.
	 * @param i     The number the rendered element is of its parent.
	 *
	 * @returns {ReactElement | string} The sidebar element to render.
	 */
	sidebar( props: RenderEditProps, i: number ): ReactElement {
		return null;
	}
	/* eslint-enable @typescript-eslint/no-unused-vars */

	/**
	 * Returns the configuration of this instruction.
	 *
	 * @returns The block configuration.
	 */
	configuration(): Partial<BlockConfiguration> {
		return {
			attributes: {
				[ this.options.name ]: {
					required: this.options.required === true,
				},
			},
		};
	}

	/**
	 * Checks if the instruction block is valid.
	 *
	 * @param blockInstance The attributes from the block.
	 *
	 * @returns {BlockValidationResult} The validation result.
	 */
	validate( blockInstance: BlockInstance ): BlockValidationResult {
		return new BlockValidationResult( blockInstance.clientId, this.constructor.name, BlockValidation.Unknown, BlockPresence.Unknown );
	}
}
