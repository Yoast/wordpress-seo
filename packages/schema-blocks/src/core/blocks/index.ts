import BlockDefinition from "./BlockDefinition";
import BlockInstruction from "./BlockInstruction";
import { BlockInstructionClass } from "./BlockInstructionClass";
import BlockLeaf from "./BlockLeaf";
import * as BlockDefinitionRepository from "./BlockDefinitionRepository";

export {
	BlockDefinition,
	BlockInstruction,
	BlockInstructionClass,
	BlockLeaf,
	BlockDefinitionRepository,
};
