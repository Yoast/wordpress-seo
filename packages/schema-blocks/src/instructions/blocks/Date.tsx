import { BlockConfiguration, BlockInstance } from "@wordpress/blocks";
import { DateTimePicker, Dropdown } from "@wordpress/components";
import { createElement, useState } from "@wordpress/element";
import { __experimentalGetSettings, dateI18n, format } from "@wordpress/date";
import { __ } from "@wordpress/i18n";
import BlockInstruction from "../../core/blocks/BlockInstruction";
import { RenderEditProps, RenderSaveProps } from "../../core/blocks/BlockDefinition";
import { useCallback } from "react";
import { BlockValidationResult } from "../..";
import { getPresence } from "../../functions/validators/getPresence";
import { BlockPresence, BlockValidation } from "../../core/validation";

/**
 * Adds a date picker to the schema block.
 */
export default class Date extends BlockInstruction {
	options: {
		name: string;
		required?: boolean;
	};

	/**
	 * The React components to show in the editor when editing this block.
	 *
	 * @param props The block's properties.
	 *
	 * @return The React components to show in the editor when editing this block.
	 */
	edit( props: RenderEditProps ): JSX.Element {
		const { attributes, setAttributes } = props;

		const dateFormat = Date.getDateFormat();

		const currentlySelectedDate = dateI18n( dateFormat, attributes[ this.options.name ] as string, false );

		const [ selectedDate, setSelectedDate ] = useState( currentlySelectedDate );

		let currentValue = __( "Select a date", "yoast-schema-blocks" );
		if ( attributes[ this.options.name ] ) {
			currentValue = format( "Y-m-d", attributes[ this.options.name ] as string );
		}

		/**
		 * Sets the selected date.
		 *
		 * @param dateTime The selected date and time in the form 'yyyy-MM-ddThh:mm:ss' (only the date part is used).
		 */
		const setDate = useCallback( ( dateTime: string ) => {
			const date = dateTime ? dateTime.split( "T" )[ 0 ] : null;

			setAttributes( {
				[ this.options.name ]: date,
			} );
			setSelectedDate( dateI18n( dateFormat, date, false ) );
		}, [ props, dateFormat, setSelectedDate ] );

		/**
		 * Render toggle.
		 *
		 * @param renderProps The render props.
		 *
		 * @return The rendered toggle element.
		 */
		const renderToggle = useCallback( ( renderProps: Dropdown.RenderProps ): JSX.Element => {
			return <button
				onClick={ renderProps.onToggle }
				aria-expanded={ renderProps.isOpen }
			>
				{ currentValue }
			</button>;
		}, [ selectedDate ] );

		/**
		 * Renders the content of the dropdown element.
		 *
		 * @returns The rendered content of the dropdown element.
		 */
		const renderContent = useCallback( (): JSX.Element => {
			return <div className="yoast-block-date-picker">
				<DateTimePicker
					currentDate={ attributes[ this.options.name ] ? selectedDate : null }
					onChange={ setDate }
				/>
			</div>;
		}, [ selectedDate, setDate ] );

		return <Dropdown
			className="yoast-block-date-picker-container"
			position="bottom center"
			renderToggle={ renderToggle }
			renderContent={ renderContent }
		/>;
	}

	/**
	 * Get the date format as set in the WordPress settings.
	 *
	 * @return The date format.
	 */
	static getDateFormat(): string {
		/**
		 * The `getSettings` function has been marked as experimental, but
		 * no alternative (REST-endpoint, setting it on a Redux store)
		 * seems to be implemented yet.
		 *
		 * @see https://github.com/WordPress/gutenberg/pull/7996/files#r224152849
		 */
		const settings = __experimentalGetSettings();
		return settings.formats.date;
	}

	/**
	 * The HTML to save to the database, as a React element.
	 *
	 * @param props The properties.
	 *
	 * @return The HTML to save to the database.
	 */
	save( props: RenderSaveProps ): JSX.Element {
		const date = props.attributes[ this.options.name ] as string;

		if ( ! date ) {
			return null;
		}

		const dateFormat = Date.getDateFormat();

		return <time dateTime={ date }>{ dateI18n( dateFormat, date, false ) }</time>;
	}

	/**
	 * The partial configuration to add to the block instruction
	 * that contains this instruction.
	 *
	 * @return The partial configuration.
	 */
	configuration(): Partial<BlockConfiguration> {
		return {
			attributes: {
				[ this.options.name ]: {
					type: "string",
				},
			},
		};
	}

	/**
	 * Checks if the instruction block is valid.
	 *
	 * @param blockInstance The attributes from the block.
	 *
	 * @returns {BlockValidationResult} The validation result.
	 */
	validate( blockInstance: BlockInstance ): BlockValidationResult {
		const date = blockInstance.attributes[ this.options.name ] as string;
		const presence = getPresence( this.options );

		let validation = BlockValidation.Unknown;
		if ( date && date.trim().length > 0 ) {
			validation = BlockValidation.Valid;
		} else {
			if ( presence === BlockPresence.Required ) {
				validation = BlockValidation.MissingRequiredAttribute;
			} else {
				validation = BlockValidation.MissingRecommendedAttribute;
			}
		}

		return new BlockValidationResult( blockInstance.clientId, this.constructor.name, validation, presence );
	}
}

BlockInstruction.register( "date", Date );
