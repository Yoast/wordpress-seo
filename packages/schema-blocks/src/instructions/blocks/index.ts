import "./Block";
import "./ClassName";
import "./CurrencySelect";
import "./Date";
import "./Heading";
import "./InheritSidebar";
import "./InnerBlocks";
import "./RichText";
import "./Select";
import "./SidebarCheckbox";
import "./SidebarDuration";
import "./SidebarInput";
import "./SidebarSelect";
import "./TextInput";
import "./Title";
import "./Variation";
import "./VariationPicker";
import "./Duration";
