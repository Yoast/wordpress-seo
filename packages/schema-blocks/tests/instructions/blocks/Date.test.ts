import * as renderer from "react-test-renderer";
import { mocked } from "ts-jest/utils";

import { __experimentalGetSettings, dateI18n, DateSettings } from "@wordpress/date";

import Date from "../../../src/instructions/blocks/Date";
import { RenderEditProps, RenderSaveProps } from "../../../src/core/blocks/BlockDefinition";

jest.mock( "@wordpress/date", () => ( {
	__experimentalGetSettings: jest.fn(),
	dateI18n: jest.fn(),
} ) );
jest.mock( "../../../src/instructions/blocks/InnerBlocks", () => ( {} ) );

const mockedGetSettings = mocked( __experimentalGetSettings, false );
const mockedDateI18n = mocked( dateI18n, false );

describe( "The Date instruction", () => {
	/**
	 * Skipped because the `useState` hook does not work with the test renderer.
	 */
	it.skip( "returns the React components for in the editor correctly.", () => {
		const options = {
			name: "publishedDate",
		};
		const dateInstruction = new Date( 12, options );

		const setAttributes = jest.fn();

		const props: RenderEditProps = {
			clientId: "1234-abc",
			className: "",
			isSelected: false,
			setAttributes,
			attributes: {
				publishedDate: "2020-21-01",
			},
		};

		const tree = renderer
			.create( dateInstruction.edit( props ) )
			.toJSON();

		expect( tree ).toMatchSnapshot();
	} );

	it( "returns the correct configuration to add to the schema block.", () => {
		const options = {
			name: "publishedDate",
		};

		const dateInstruction = new Date( 12, options );

		const expectedConfiguration = {
			attributes: {
				publishedDate: {
					type: "string",
				},
			},
		};

		const actualConfiguration = dateInstruction.configuration();

		expect( actualConfiguration ).toEqual( expectedConfiguration );
	} );

	it( "returns the correct React components to render when the post is saved.", () => {
		const options = {
			name: "publishedDate",
		};

		const dateInstruction = new Date( 12, options );

		const props: RenderSaveProps = {
			clientId: "1234-abc",
			attributes: {
				publishedDate: "2020-21-01",
			},
		};

		const dateSettings = {
			formats: {
				date: "y-m-d",
			},
		} as DateSettings;

		mockedGetSettings.mockReturnValue( dateSettings );

		mockedDateI18n.mockReturnValue( "2020-21-01" );

		const tree = renderer.create( dateInstruction.save( props ) ).toJSON();

		expect( tree ).toMatchSnapshot();
	} );
} );
