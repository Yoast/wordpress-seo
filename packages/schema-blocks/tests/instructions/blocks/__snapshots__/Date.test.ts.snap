// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Date instruction returns the correct React components to render when the post is saved. 1`] = `
<time
  dateTime="2020-21-01"
>
  2020-21-01
</time>
`;
